elkscmd/man/man1/df.1
.TH DF 1
.SH NAME
df \- report on free disk space
.SH SYNOPSIS
\fBdf\fP [\fB\-ikP\fP] [\fB\-t\fP \fItype\fP] [\fIfile\fP ...]
.SH DESCRIPTION
.B Df
lists the amount of free space on the currently mounted devices (no arguments),
or the device given as an argument.  If the argument is not a device buf a file or a directory, then the
device it resides on is listed. FAT file systems are supported if the result would 
be meaningful. I.e. not when using the \fB-i\fP or \fB-P\fP options.
.PP
File systems are always displayed in the order they were mounted.
.SS OPTIONS
Without options,
.B df
will give a listing like this:
.sp
.nf
Filesystem    1k-Blocks     free     used    %  FUsed%  Mounted on
/dev/hda3         65535    61443     4092   7%     6%   /
/dev/fd0           1440      380     1060  74%    73%   /fd
/dev/hda1        348226   345512     2714   1%          /mnt (FAT)
/dev/hdb1           726      554      172  24%          /mnt5 (FAT)
/dev/hda2         65535    59812     5723   9%     2%   /mnt3
/dev/hda4         41328    37129     4199  11%     3%   /mnt4
.fi
.PP
The
.B \-i
option shifts the focus to inodes:
.sp
.nf
Filesystem       Inodes     IUsed    IFree    %IUsed   Mounted on
/dev/hda3         21845      2298    19547       11%   /
/dev/hda1         [Not a MINIX filesystem]             /mnt
/dev/hda2         21845       228    21617        2%   /mnt2
/dev/hda4         13776       313    13463        3%   /mnt4
/dev/hdb1         [Not a MINIX filesystem]             /mnt5
.fi
.PP
Option
.B \-P
makes
.B df
use \s-2POSIX\s+2 defined output in 512 byte units:
.sp
.nf
Filesystem     512-blocks    Used Available Capacity Mounted on
/dev/hda3        131070     13876    117194      11% /
/dev/hda1         [Not a MINIX filesystem]           /mnt
/dev/hda2        131070     11446    119624       9% /mnt2
/dev/hda4         82656      8398     74258      11% /mnt4
/dev/hdb1         [Not a MINIX filesystem]           /mnt5
.fi
.PP
With
.B \-k
forces the use of 1024 byte units with the -P option.
.SH BUGS
Default output should also be in 512 byte units says \s-2POSIX\s+2.
Getting data for FAT filesystems is slow. To just get a fast listing of mounted file systems, 
use the \fBmount\fP command without parameters.
.SH AUTHOR
Kees J. Bot (kjb@cs.vu.nl)
.SH "SEE ALSO"
.BR du (1)
.BR mount (8).


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/file_utils/df.c
/* df - disk free block printout	Author: Andy Tanenbaum
 *
 * 91/04/30 Kees J. Bot (kjb@cs.vu.nl)
 *	Map filename arguments to the devices they live on.
 *	Changed output to show percentages.
 *
 * 92/12/12 Kees J. Bot
 *	Posixized.  (Almost, the normal output is in kilobytes, it should
 *	be 512-byte units.  'df -P' and 'df -kP' are as it should be.)
 *
 * Ported to ELKS by Greg Haerr May 2020
 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <limits.h>
#include <fcntl.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/mount.h>
#include <linuxmt/minix_fs.h>
#include <linuxmt/fs.h>

#define BLOCK_SIZE	1024

int df(char *device, char *mpnt);

typedef unsigned int bit_t;
bit_t bit_count(unsigned blocks, bit_t bits, int fd);
struct dnames {
	char *name;
	char *mpoint;
};
struct dnames *devname(char *);
static char *dev_name(dev_t);
int df_fat(char *, int);

int iflag= 0;	/* Focus on inodes instead of blocks. */
int Pflag= 0;	/* Posix standard output. */
int kflag= 0;	/* Output in kilobytes instead of 512 byte units for -P. */
int istty;	/* isatty(1) */

/* (num / tot) in percentages rounded up. */
#define percent(num, tot)  ((int) ((100L * (num) + ((tot) - 1)) / (tot)))

/* One must be careful printing all these _t types. */
#define L(n)	((long) (n))

void usage(void)
{
	fprintf(stderr, "usage: df [-ikP] [device_or_mount_point]\n");
	exit(1);
}

int main(int argc, char *argv[])
{
  char *device = "/";
  char *blockdev;
  struct dnames *dname;
  struct statfs statfs;

  while (argc > 1 && argv[1][0] == '-') {
  	char *opt= argv[1]+1;

  	while (*opt != 0) {
  		switch (*opt++) {
  		case 'i':	iflag= 1;	break;
  		case 'k':	kflag= 1;	break;
  		case 'P':	Pflag= 1;	break;
		default:	usage();
		}
	}
	argc--;
	argv++;
  }

  istty= isatty(1);

  if (argc > 1)
	device = argv[1];

  dname = devname(device);
  if (!(blockdev = dname->name)) {
	fprintf(stderr, "Can't find /dev/ device for %s\n", device);
	exit(1);
  }

  if (iflag) {
	printf("Filesystem       Inodes     IUsed    IFree    %%IUsed   Mounted on\n");
  } else {
	if (!Pflag) 
		printf("Filesystem    1K-blocks     Free     Used    %%  FUsed%%  Mounted on\n");
	else 
		printf("Filesystem    %4d-blocks    Used Available Capacity Mounted on\n",
			kflag ? 1024 : 512);
  }

  if (argc == 1) {	/* loop through all mounted devices */
	int i;
  	for (i = 0; i < NR_SUPER; i++) {
		if (ustatfs(i, &statfs, UF_NOFREESPACE) >= 0) {
			char *nm = dev_name(statfs.f_dev);
			char *filler = "  ";		/* Text alignment */
			if (Pflag) filler = "";
			if (statfs.f_type > FST_MSDOS || (statfs.f_type == FST_MSDOS && (Pflag||iflag))) 
				printf("%-17s %-34s %s%s\n", nm, "[Not a MINIX filesystem]", filler, statfs.f_mntonname);
			else if (statfs.f_type == FST_MSDOS && !Pflag)
				df_fat(nm, i);
			else
				df(nm, statfs.f_mntonname);
		}
	}
	return 0;
  } else
  	return df(blockdev, dname->mpoint);
}

int df_fat(char *name, int dev) 
{
	struct statfs stat;
	if (ustatfs(dev, &stat, 0) < 0)
		return -1;
	printf("%-15s %7ld  %7ld  %7ld %3d%%          %s (FAT)\n", name,
	    stat.f_blocks, stat.f_bfree, stat.f_blocks - stat.f_bfree,
	    percent(stat.f_blocks - stat.f_bfree, stat.f_blocks), stat.f_mntonname);
	return 0;
}

int df(char *device, char *mpnt)
{
  int fd;
  bit_t i_count, z_count;
  long totblocks, busyblocks;
  int n;
  struct minix_super_block super, *sp;

  if ((fd = open(device, O_RDONLY)) < 0) {
	fprintf(stderr, "df: %s: %s\n", device, strerror(errno));
	return(1);
  }
  lseek(fd, (off_t) BLOCK_SIZE, SEEK_SET);	/* skip boot block */

  if (read(fd, (char *) &super, sizeof(super)) != (int) sizeof(super)) {
	fprintf(stderr, "df: Can't read super block of %s\n", device);
	close(fd);
	return(1);
  }

  lseek(fd, (off_t) BLOCK_SIZE * 2L, SEEK_SET);	/* skip rest of super block */
  sp = &super;
  if (sp->s_magic != MINIX_SUPER_MAGIC && sp->s_magic != MINIX_SUPER_MAGIC2) {
	fprintf(stderr, "df: %s: Not a MINIX file system\n", device);
	close(fd);
	return(1);
  }
  if (sp->s_magic == MINIX_SUPER_MAGIC) sp->s_zones= sp->s_nzones;

  i_count = bit_count(sp->s_imap_blocks, (bit_t) (sp->s_ninodes+1), fd);

  if (i_count == -1) {
	fprintf(stderr, "df: Can't find bit maps of %s\n", device);
	close(fd);
	return(1);
  }
  i_count--;	/* There is no inode 0. */

  /* The first bit in the zone map corresponds with zone s_firstdatazone - 1
   * This means that there are s_zones - (s_firstdatazone - 1) bits in the map
   */
  z_count = bit_count(sp->s_zmap_blocks,
	(bit_t) (sp->s_zones - (sp->s_firstdatazone - 1)), fd);

  if (z_count == -1) {
	fprintf(stderr, "df: Can't find bit maps of %s\n", device);
	close(fd);
	return(1);
  }
  /* Don't forget those zones before sp->s_firstdatazone - 1 */
  z_count += sp->s_firstdatazone - 1;

  totblocks = (block_t) sp->s_zones << sp->s_log_zone_size;
  busyblocks = (block_t) z_count << sp->s_log_zone_size;

  /* Print results. */
  printf("%s", device);
  n = strlen(device);
  if (n > 15 && istty) { printf("\n"); n= 0; }
  while (n < 15) { printf(" "); n++; }

  if (iflag) {
	printf(" %7ld   %7ld  %7ld     %4d%%   %s\n",
		L(sp->s_ninodes),			/* Inodes */
		L(i_count),				/* IUsed */
		L(sp->s_ninodes - i_count),		/* IAvail */
		percent(i_count, sp->s_ninodes),	/* Capacity */
		mpnt					/* Mount pnt */

	);
  }
  if (!Pflag && !iflag) {
	printf(" %7ld  %7ld  %7ld %3d%%   %3d%%   %s\n",
		L(totblocks),				/* Blocks */
		L(totblocks - busyblocks),		/* free */
		L(busyblocks),				/* used */
		percent(busyblocks, totblocks),		/* % */
		percent(i_count, sp->s_ninodes),	/* FUsed% */
		mpnt					/* Mount point */
	);
  }
  if (Pflag && !iflag) {
  	if (!kflag) {
  		/* 512-byte units please. */
  		totblocks *= 2;
  		busyblocks *= 2;
	}
	printf(" %7ld   %7ld   %7ld    %4d%% %s\n",
		L(totblocks),				/* Blocks */
		L(busyblocks),				/* Used */
		totblocks - busyblocks,			/* Available */
		percent(busyblocks, totblocks),		/* Capacity */
		mpnt					/* Mount point */
	);
  }
  close(fd);
  return(0);
}

bit_t bit_count(unsigned blocks, bit_t bits, int fd)
{
  char *wptr;
  int i, b;
  bit_t busy;
  char *wlim;
  static char buf[BLOCK_SIZE];
  static char bits_in_char[1 << CHAR_BIT];

  /* Precalculate bitcount for each char. */
  if (bits_in_char[1] != 1) {
	for (b = (1 << 0); b < (1 << CHAR_BIT); b <<= 1)
		for (i = 0; i < (1 << CHAR_BIT); i++)
			if (i & b) bits_in_char[i]++;
  }

  /* Loop on blocks, reading one at a time and counting bits. */
  busy = 0;
  for (i = 0; i < blocks && bits != 0; i++) {
	if (read(fd, buf, BLOCK_SIZE) != BLOCK_SIZE) return(-1);

	wptr = &buf[0];
	if (bits >= CHAR_BIT * BLOCK_SIZE) {
		wlim = &buf[BLOCK_SIZE];
		bits -= CHAR_BIT * BLOCK_SIZE;
	} else {
		b = bits / CHAR_BIT;	/* whole chars in map */
		wlim = &buf[b];
		bits -= b * CHAR_BIT;	/* bits in last char, if any */
		b = *wlim & ((1 << bits) - 1);	/* bit pattern from last ch */
		busy += bits_in_char[b];
		bits = 0;
	}

	/* Loop on the chars of a block. */
	while (wptr != wlim)
		busy += bits_in_char[*wptr++ & ((1 << CHAR_BIT) - 1)];
  }
  return(busy);
}

/*
 * $PchId: df.c,v 1.7 1998/07/27 18:42:17 philip Exp $
 */

/* return /dev/ device from dirname */
struct dnames *devname(char *dirname)
{
   static char dev[] = "/dev";
   struct stat st, dst;
   DIR  *fp;
   struct dirent *d;
   static struct statfs statfs;
   static struct dnames dn;
   static char name[MAXNAMLEN];

   if (!strncmp(dirname, "/dev/", 5)) {
	dn.name = dirname;
	dn.mpoint = dirname;
	return &dn;
   }

   if (stat(dirname, &st) < 0)
      return 0;

   fp = opendir(dev);
   if (fp == 0)
      return 0;
   strcpy(name, dev);
   strcat(name, "/");

   while ((d = readdir(fp)) != 0) {
      if(d->d_name[0] == '.' || strlen(d->d_name) > sizeof(name) - sizeof(dev) - 1)
         continue;
      strcpy(name + sizeof(dev), d->d_name);
      if (stat(name, &dst) == 0) {
	 if (st.st_dev == dst.st_rdev) {
	     if (ustatfs(st.st_dev, &statfs, UF_NOFREESPACE) < 0) {
		dn.mpoint = NULL;
	     } else {
		dn.mpoint = statfs.f_mntonname;
	     }
	     closedir(fp);
	     dn.name = name;
	     return &dn;
        }
      }
   }
   closedir(fp);
   return NULL;
}

/*
 * Convert a block device number to name.
 * From mount.c
 */
static struct dev_name_struct {
        char *name;
        dev_t num;
} devices[] = {
        /* root_dev_name needs first 5 in order*/
        { "hda",     0x0300 },
        { "hdb",     0x0320 },
        { "hdc",     0x0340 },
        { "hdd",     0x0360 },
        { "fd0",     0x0380 },
        { "fd1",     0x03a0 },
        { "ssd",     0x0200 },
        { "rd",      0x0100 },
        { NULL,           0 }
};

static char *dev_name(dev_t dev)
{
	int i;
#define NAMEOFF 5
	static char name[10] = "/dev/";

	for (i=0; i<sizeof(devices)/sizeof(devices[0])-1; i++) {
		if (devices[i].num == (dev & 0xfff0)) {
			strcpy(&name[NAMEOFF], devices[i].name);
			if (i < 4) {
				if (dev & 0x07) {
					name[NAMEOFF+3] = '0' + (dev & 7);
					name[NAMEOFF+4] = 0;
				}
			}
			return name;
		}
	}
	return NULL;
}
